lambda/enable_guardduty.py

```
@@ -0,0 +1,330 @@
+#!/usr/bin/env python3
+
+import boto3
+from botocore.exceptions import ClientError
+import json
+import logging
+import os
+import time
+
+
+logger = logging.getLogger()
+logger.setLevel(logging.INFO)
+# Quiet Boto3
+logging.getLogger('botocore').setLevel(logging.WARNING)
+logging.getLogger('boto3').setLevel(logging.WARNING)
+
+# Deploy Guard Duty across all child accounts to the payer account
+# Process documented here:
+# https://docs.aws.amazon.com/guardduty/latest/ug/guardduty_accounts.html#guardduty_become_api
+
+
+def handler(event, context):
+    '''
+    message = {
+        'account_id': 'string',
+        'message': 'string',  // optional, if un-specified, ENV var (from CFT)
+        'dry_run': true|false,  // optional, if un-specified, dry_run=false
+        'region': ['string'],  // optional, if un-specified, runs all regions
+    }
+    '''
+    logger.info("Received event: " + json.dumps(event, sort_keys=True))
+    message = json.loads(event['Records'][0]['Sns']['Message'])
+    logger.info("Received message: " + json.dumps(message, sort_keys=True))
+
+    if "account_id" not in message:
+        error_message = "message['account_id'] must be specified"
+        logger.error(error_message)
+        raise KeyError(error_message)
+
+    # get parent organization's account_id
+    role_arn = create_role_arn(message['account_id'], os.environ['ASSUME_ROLE'])
+    creds = get_creds(role_arn)
+    if creds is False:
+        logger.error(f"Unable to assume role {role_arn} to get parent organization")
+        return
```

This should also raise an error and cause the Lambda execution to error. return() ends up as a normal execution.